File: 프로그래머스/카카오/[1차]캐시.java
import java.util.HashMap;
import java.util.Map;
import java.util.Locale;

class Solution {
     public int solution(int cacheSize, String[] cities) {
            int answer = 0;
            Map<String, Integer> ranking = new HashMap<>(); // 도시이름, 순위 (1~cacheSize만큼 우선순위)

            for (int i = 0; i < cities.length; i++) {
                cities[i] = cities[i].toLowerCase(Locale.ROOT);
            }

            int level = 0;
            if (cacheSize == 0) {
                for (int i = 0; i < cities.length; i++) {
                    answer += 5;
                }
            } else {
                for (String s : cities) {
                    if (ranking.containsKey(s)) {
                        ranking.put(s, level++);
                        answer++;
                        continue;
                    }
                    if (ranking.size() == cacheSize) {
                        int min = 100001;
                        String city = "";
                        for (var key : ranking.keySet()) {
                            if (ranking.get(key) < min) {
                                city = key;
                                min = ranking.get(key);
                            }
                        }
                        ranking.remove(city);
                    }
                    ranking.put(s, level);
                    answer += 5;
                    level++;
                }
            }

            return answer;
        }
}

File: 프로그래머스/카카오/[1차] 캐시.py
import operator


def solution(cacheSize, cities):
    answer = 0
    mp = dict()
    if cacheSize == 0:
        return len(cities)*5
    for id in range(0, len(cities)):
        city = "".join(cities[id]).lower()
        if city in mp:  # 캐시에 저장되어있을때
            mp[city] = id
            answer += 1
        else:
            if len(mp) < cacheSize:
                mp[city] = id
            else:
                s = min(mp.items(), key=operator.itemgetter(1))
                del mp[s[0]]
                mp[city] = id
            answer += 5
    return answer
